modules/core/src/smithy4s/Blob.scala

```diff
@@ -41,78 +117,171 @@ object Blob {
+    override def asByteBufferUnsafe(start: Int, size: Int): ByteBuffer = {
+      val b = ByteBuffer.wrap(arr, start.toInt, size)
+      if (start == 0 && size == arr.length) b
+      else b.slice()
+    }
+
+    override def copyToArray(xs: Array[Byte], start: Int, offset: Int, size: Int): Unit =
+      System.arraycopy(arr, offset.toInt, xs, start, size)
+
+    override def copyToStream(s: OutputStream, offset: Int, size: Int): Unit =
+      s.write(arr, offset.toInt, size.toInt)
+
+    override def copyToBuffer(buffer: ByteBuffer, offset: Int, size: Int): Int = {
+      val toCopy = buffer.remaining.min(size)
+      buffer.put(arr, offset.toInt, toCopy)
+      toCopy
     }
 
-    override def sameBytesAs(other: Blob): Boolean = other match {
-      case otherBlob: ByteArrayBlob =>
-        java.util.Arrays.equals(bytes, otherBlob.bytes)
-      case otherBlob: ByteBufferBlob =>
-        ByteBuffer.wrap(bytes).compareTo(otherBlob.buffer) == 0
+    override def toArray: Array[Byte] = {
+      val newArray = Array.ofDim[Byte](arr.length)
+      arr.copyToArray(newArray, 0, arr.length)
+      newArray
+    }
+    override def toArrayUnsafe: Array[Byte] = arr
+
+    override def isEmpty: Boolean = arr.isEmpty
+    override def size: Int = arr.length
+
+    override def toString = {
+      s"ByteArrayBlob[${Base64.getEncoder().encodeToString(arr)}]"
     }
 
     override def equals(other: Any): Boolean = {
       other.isInstanceOf[ByteArrayBlob] &&
-      java.util.Arrays.equals(bytes, other.asInstanceOf[ByteArrayBlob].bytes)
+      java.util.Arrays.equals(arr, other.asInstanceOf[ByteArrayBlob].arr)
     }
 
     override def hashCode(): Int = {
       var hashCode = 0
       var i = 0
-      while (i < bytes.length) {
-        hashCode += bytes(i).hashCode()
+      while (i < arr.length) {
+        hashCode += arr(i).hashCode()
         i += 1
       }
       hashCode
     }
   }
-  private final class ByteBufferBlob(val buffer: ByteBuffer) extends Blob {
-    override def toString = s"ByteBufferBlob[${buffer.toString()}]"
-    override def isEmpty: Boolean = !buffer.hasRemaining()
-    override def size: Int = buffer.remaining()
-    override def hashCode = buffer.hashCode()
-    override def sameBytesAs(other: Blob): Boolean = other match {
-      case otherBlob: ByteBufferBlob =>
-        buffer.compareTo(otherBlob.buffer) == 0
-      case otherBlob: ByteArrayBlob =>
-        buffer.compareTo(ByteBuffer.wrap(otherBlob.toArray)) == 0
+  final class ByteBufferBlob(val buf: ByteBuffer) extends Blob {
+    def apply(i: Int) = buf.get(i.toInt)
+
+    override def copyToArray(xs: Array[Byte], start: Int, offset: Int, size: Int): Unit = {
+      val n = buf.duplicate()
+      n.position(offset.toInt)
+      n.get(xs, start, size)
+      ()
+    }
+
+    override def toArray: Array[Byte] = {
+      val arr = Array.ofDim[Byte](buf.remaining())
+      copyToArray(arr, 0, 0, size)
+      arr
+    }
+
```

Comment:
Many `ByteBuffer`s are actually backed by an `Array[Byte]` so we can override `toArrayUnsafe` (and `toArraySliceBlob`) with something more efficient.

https://github.com/typelevel/fs2/blob/196995d58d5b82da38252ec81d003f91a740503a/core/shared/src/main/scala/fs2/Chunk.scala#L1026-L1031